Add certificate timestamp validation

Captain-P-Goldfish opened the pull request:
User-certificates that are used for mutual client authentication can
now be rejected if the certificate is not valid yed or has already
been expired

Captain-P-Goldfish committed:
Add certificate timestamp validation

User-certificates that are used for mutual client authentication can
now be rejected if the certificate is not valid yed or has already
been expired

mposolda commented:
@Captain-P-Goldfish Could you please create JIRA and describe in more details what is the issue, which you're trying to solve? Thanks

Captain-P-Goldfish commented:
done
https://issues.jboss.org/browse/KEYCLOAK-11818

stianst commented:
Could you send a mail on the developer mailing list around this change so we can get feedback from others?

stianst commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -468,6 +469,24 @@ public CertificateValidator validateExtendedKeyUsage() throws GeneralSecurityExc
+        X509Certificate x509Certificate = _certChain[_certChain.length - 1];
+        // this might create time-offset issues but those will probably be negligible for certificates should
+        // be renewed and exchanged in due time
+        if (x509Certificate.getNotBefore().getTime() > Instant.now().toEpochMilli()) {
```
Should use org.keycloak.common.util.Time as that supports tweaking the time for testing purposes

Captain-P-Goldfish replied:
done

stianst commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -468,6 +469,24 @@ public CertificateValidator validateExtendedKeyUsage() throws GeneralSecurityExc
+            String message = "certificate is not valid yet: " + x509Certificate.getNotBefore().toInstant().toString();
+            throw new GeneralSecurityException(message);
+        }
+        if (x509Certificate.getNotAfter().getTime() < Instant.now().toEpochMilli()) {
```
Should use org.keycloak.common.util.Time as that supports tweaking the time for testing purposes

Captain-P-Goldfish replied:
done

stianst reviewed (changes requested):
Use org.keycloak.common.util.Time to support testing, and add tests.

Captain-P-Goldfish committed:
Add unit tests for certificate validation

stianst commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -468,6 +469,22 @@ public CertificateValidator validateExtendedKeyUsage() throws GeneralSecurityExc
         return this;
     }
 
+    public CertificateValidator validateTimestamps(boolean isValidationEnabled) throws GeneralSecurityException {
```
Should we not validate the not-before and not-after on all certs in the chain rather than just the last one?

Captain-P-Goldfish replied:
not sure if this might cause other problems in some PKI's. That is why I just validated the last one.

Captain-P-Goldfish replied:
done

stianst commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/X509AuthenticatorConfigModel.java:
```diff
@@ -253,6 +253,15 @@ public X509AuthenticatorConfigModel setCanonicalDnEnabled(boolean value) {
     }
+
+    public boolean isCertValidationEnabled() {
+        return Boolean.parseBoolean(getConfig().get(TIMESTAMP_VALIDATION));
```
For existing users that upgrade to a new release TIMESTAMP_VALIDATION is not set, so this should make sure it works when getConfig().get(TIMESTAMP_VALIDATION) is null

Captain-P-Goldfish replied:
done

stianst commented on services/src/test/java/org/keycloak/authentication/authenticators/x509/CertificateValidatorTest.java:
```diff
@@ -0,0 +1,157 @@
+package org.keycloak.authentication.authenticators.x509;
```
I would much prefer if this was converted to a functional test by extending one of the tests in testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509. Would that be achievable?

Captain-P-Goldfish replied:
Tried it but unfortunately I got trouble in getting it running. Could not determine what exactly the problem was and how to debug. Currently I am not having enough time to check this out in detail...

Captain-P-Goldfish replied:
was not able to get the test running and I am not getting any answers in the mailing list...
https://groups.google.com/forum/#!topic/keycloak-dev/MI20c_zKI1Q

Captain-P-Goldfish committed:
Add unit test for timestamp validation

Assert that no exception occurs if the new timestamp
validation has not been previously set and is still null

Captain-P-Goldfish committed:
Validate all certificates in the chain

Captain-P-Goldfish committed:
Merge branch 'certificate-validity' of https://github.com/Captain-P-Goldfish/keycloak into certificate-validity

slaskawi commented on services/src/test/java/org/keycloak/authentication/authenticators/x509/CertificateValidatorTest.java:
```diff
@@ -0,0 +1,157 @@
+      validator.validateTimestamps(true);
+    } catch (Exception ex)
+    {
+      ex.printStackTrace();
```
nit: you can just remove the catch block and the end result will be the same.

slaskawi reviewed (dismissed):
Thanks for the Pull Request @Captain-P-Goldfish !

The code itself looks good to me. I believe you use some non-default IDE settings that put `{` in separate lines (as you probably noticed from the other parts of the codebase, we don't do that). I think it would be good to align this to the rest of the codebase.

mposolda reviewed (changes requested):
Thanks for the review!

@Captain-P-Goldfish I will approve once you fix the formatting - the brackets will be ideal on same line and the padding should be 4 characters instead of 2. See the rest of the codebase for more details. Thanks

Captain-P-Goldfish committed:
Adjust code formatting

slaskawi reviewed (approved):
LGTM! @stianst @hmlnarik @mposolda This one is ready to be merged.

mposolda reviewed (approved):
@Captain-P-Goldfish @slaskawi Thanks!
